### PR TITLE
fix(acm): Allow tokens as a part of the hosted zone name

### LINKED ISSUE #6133: DnsValidatedCertificate: Tokens in the hostedZone property breaks cdk synth
## 🐛 Bug Report

For the [DnsValidatedCertificate](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-certificatemanager.DnsValidatedCertificate.html) construct, [tokens](https://docs.aws.amazon.com/cdk/latest/guide/tokens.html) within the `hostedZone` property break the `cdk synth` with the following error:

```
DNS zone ${Token[Default.Parameter.Value.1234]} is not authoritative for certificate domain name mydomain.com
```

Ideally, the [validation](https://github.com/aws/aws-cdk/blob/e3305d8d09b3fd87a0609606c3638332c4eeef33/packages/%40aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts#L94-L101) done in the construct should take into account tokens and not try to validate the string value of the token since, well, that is useless since it will always end up being something like `${Token[Default.Parameter.Value.1371]}` opposed to the domain you would expect.

## Background

We have our Route53 hosted zone deployed in `us-east-1`. However, we run a global service and need to be able to reference the hosted zone from any region to be able to do things like creating certificates, creating region-specific CNAMES, etc.

With that and to get past the inability to import values directly, we write the hosted zone id and name into SSM's parameter store using the [StringParameter](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ssm.StringParameter.html) construct.

Then, whenever we need the `IHostedZone` reference, we leverage a custom resource (`SSMParameterReader`) that enables us to read parameters stored in whatever region we specify.

```typescript
interface SSMParameterReaderProps {
  parameterName: string;
  region: string;
}

/**
 * A custom construct that is an AwsCustomResource that will use the provided parameterName and region
 * to make a call to SSM to retrieve the parameter's value in that region.
 *
 * This can be used alongside @aws-cdk/aws-ssm.StringParameter to reference values across regions
 */
export class SSMParameterReader extends AwsCustomResource {
  constructor(scope: Construct, name: string, props: SSMParameterReaderProps) {
    const { parameterName, region } = props;

    const ssmAwsSdkCall: AwsSdkCall = {
      service: 'SSM',
      action: 'getParameter',
      parameters: {
        Name: parameterName
      },
      region,
      physicalResourceId: Date.now().toString() // Update physical id to always fetch the latest version
    };

    super(scope, name, { onUpdate: ssmAwsSdkCall });
  }

  /**
   * Returns the string value of the parameter stored in the SSM parameter store
   */
  public getParameterValue(): string {
    return this.getData('Parameter.Value').toString();
  }
}

```

Using that custom resource, we can then read the hosted zone id/name and [HostedZone.fromHostedZoneAttributes](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-route53.HostedZone.html#static-from-wbr-hosted-wbr-zone-wbr-attributesscope-id-attrs) method.

```typescript
const hostedZoneIdReader = new SSMParameterReader(this, 'Route53HostedZonedIdReader', {
  parameterName: 'ROUTE_53_HOSTED_ZONE_ID',
  region: 'us-east-1'
});
const hostedZoneId: string = hostedZoneIdReader.getParameterValue();

const hostedZoneNameReader = new SSMParameterReader(this, 'Route53HostedZoneNameReader', {
  parameterName: 'ROUTE_53_HOSTED_ZONE_NAME',
  region: 'us-east-1'
});
const hostedZoneName: string = hostedZoneNameReader.getParameterValue();

const hostedZone = HostedZone.fromHostedZoneAttributes(this, 'Route53HostedZone', { hostedZoneId, zoneName: hostedZoneName });
```

With this, we can then use the `hostedZone` like it was deployed locally. We have used this extensively for creating CNAMES and aliases in the hosted zone from any region but we ran into issues whenever we tried to use the `DnsValidatedCertificate` since it's [validation logic](https://github.com/aws/aws-cdk/blob/e3305d8d09b3fd87a0609606c3638332c4eeef33/packages/%40aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts#L94-L101) doesn't take into account tokens which are what the output of custom resources get referenced as.

### Environment

  - **CDK CLI Version:** `1.20.0`
  - **Module Version:** 1.20.0
  - **OS:** macOS Mojave
  - **Language:** TypeScript (but likely all languages)

### PR DESCRIPTION
**Pull request by rhermes62**
This review adds a basic check to not try and validate unresolved tokens when performing validation for the `DnsValidatedCertificate` construct.

fixes #6133

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts [hunk 1]
```diff
@@ -109,7 +109,11 @@ export class DnsValidatedCertificate extends cdk.Resource implements ICertificat
     protected validate(): string[] {
         const errors: string[] = [];
         // Ensure the zone name is a parent zone of the certificate domain name
-        if (this.domainName !== this.normalizedZoneName && !this.domainName.endsWith('.' + this.normalizedZoneName)) {
+        if (
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts]: This indentation looks a tad weird to me... how about this?

```ts
        if (!cdk.Token.isUnresolved(this.normalizedZoneName) &&
              this.domainName !== this.normalizedZoneName &&
              !this.domainName.endsWith('.' + this.normalizedZoneName)) {
            errors.push(`DNS zone ${this.normalizedZoneName} is not authoritative for certificate domain name ${this.domainName}`);
        }
```

File: packages/@aws-cdk/aws-certificatemanager/test/test.dns-validated-certificate.ts [hunk 2]
```diff
@@ -98,6 +98,23 @@ export = {
...
+    const helloDotComZone = new PublicHostedZone(stack, 'HelloDotCom', {
+      zoneName: Token.asString('hello.com')
+    });
+
+    new DnsValidatedCertificate(stack, 'Cert', {
+      domainName: 'hello.com',
+      hostedZone: helloDotComZone
+    });
+
+    // a bit of a hack: expect(stack) will trigger validation.
```
Comment (skinny85) [inline packages/@aws-cdk/aws-certificatemanager/test/test.dns-validated-certificate.ts]: Nice hack with `expect`, but there's a simpler way :). We have a `SynthUtils` class with a `synthesize` method that we use for this kind of situations, so the test becomes:

```ts
  'does not try to validate unresolved tokens'(test: Test) {
    const stack = new Stack();

    const helloDotComZone = new PublicHostedZone(stack, 'HelloDotCom', {
      zoneName: Token.asString('hello.com')
    });

    new DnsValidatedCertificate(stack, 'Cert', {
      domainName: 'hello.com',
      hostedZone: helloDotComZone
    });

    test.doesNotThrow(() => {
      SynthUtils.synthesize(stack);
    });

    test.done();
  },
```

Comment (rhermes62) [reply packages/@aws-cdk/aws-certificatemanager/test/test.dns-validated-certificate.ts]: Perfect! Ill also update test above this that uses that hack (I didn't come up with it but just went along with it haha).

Do you think it's also worth allowing the `domainName` property in the `DnsValidatedCertificate` to be a token as well? (For me specifically, I only have/need the `zoneName` to be a `Token` but I can see others having the `domainName` in the `DnsValidatedCertificate` be a token.

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts (hunk 1 above)
Comment (rhermes62) [reply packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts]: Works with me!

File: packages/@aws-cdk/aws-certificatemanager/test/test.dns-validated-certificate.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-certificatemanager/test/test.dns-validated-certificate.ts]: I think it's fine to leave it as it is right now - we can always change it if it's needed.

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts]: Code in the current revision is still the same though 🙂

**Comment by rhermes62** [comment]
> [This comment](https://github.com/aws/aws-cdk/pull/6685#discussion_r391800075) has still not been addressed, in spite of being Resolved 🙂.

Just updated! Should show correctly now

**Pull request merged**